discv5: protocol version v5.1

fjl opened the pull request:
This change defines the first stable (non-draft) spec version of discv5. 

Work in progress. Still needs:

- [x] description of the obfuscation proposal two from #152 
- [x] test vectors

Closes #152
Closes #141 
Closes #135
Closes #131 
Closes #117 
Closes #83

pipermerriam commented:
Looking at the wire protocol document it doesn't appear to specify the computation of the `checksum` field.  It does appear to be defined in https://github.com/ethereum/devp2p/issues/152 as `crc64("discv5" || dest-id)`

pipermerriam commented:
I'm realizing that the new documents don't mention the `checksum` and it is only present in the svg images that visualize the packet encoding.

fjl commented:
Yeah, sorry, I need to update the images. In a discussion a couple weeks ago, we realized that the checksum is silly because the obfuscation layer hides away the plaintext protocol identifier anyway.

pipermerriam commented:
@fjl are you referring to "Proposal 2" from https://github.com/ethereum/devp2p/issues/152 which uses `aesctr_encrypt(masking-key, iv, plain-header)` to obfuscate the header?

fjl commented:
Yes

pipermerriam commented:
I've generated the following test vectors that I'd be interested in validating since they represent my collective interpretation of the spec as documented in this PR, #152, as well as potentially some inference in areas where neither fully documents something yet.

```
# MessagePacket (flag=0)
initiator-key: 0x01010101010101010101010101010101
nonce: 0x020202020202020202020202
source-node-id: 0x0303030303030303030303030303030303030303030303030303030303030303
dest-node-id: 0x0404040404040404040404040404040404040404040404040404040404040404
message: PING: request-id=0x01  enr-sequence-number=0x00
auth-data: MessagePacket aesgcm-nonce= 0x050505050505050505050505

encoded-packet: 0x0aa2702c8135c1276a1c00c81e04f6cec58e496f3d2dd3283f34cbdeb97364e15e201821954a5ad0f20a16d9385f1d56eb847b9d8b936c09661ee331db73fd4166474b1a962a42522611e34e608ad5f8b619eeef6ad820edf74d35

# WhoAreYouPacket (flag=1)
initiator-key: 0x01010101010101010101010101010101
nonce: 0x020202020202020202020202
source-node-id: 0x0303030303030303030303030303030303030303030303030303030303030303
dest-node-id: 0x0404040404040404040404040404040404040404040404040404040404040404
message: PING: request-id=0x01  enr-sequence-number=0x00
auth-data: WhoAreYouPacket: request-nonce=0x050505050505050505050505  id-nonce=0x0606060606060606060606060606060606060606060606060606060606060606  enr-sequence-number=0x7
encoded-packet: 0xdb05f3d612ef0bf450a932aab441ab0b794619e0818f750601eeeb25017f1be55e187c749f21dab55d3031ecf7a0395adddab0c85508c14a92cd3ecefb2536aec95c618998c914a8bf7f9998bce735cc305ae648e825eb28554b6abd79193724a72a043a3dabc6cdecb49ba9e35251522611e35ecf69051276c32e0b149b8fee2c6729

# HandshakePacket (flag-2) (WITHOUT ENR)
initiator-key: 0x01010101010101010101010101010101
nonce: 0x020202020202020202020202
source-node-id: 0x0303030303030303030303030303030303030303030303030303030303030303
dest-node-id: 0x0404040404040404040404040404040404040404040404040404040404040404
message: PING: request-id=0x01  enr-sequence-number=0x00
auth-data: HandshakePacket: 
    auth-data-head=(version=0x01 signature_size=0x40 ephemeral_key_size=0x21)  
    id-signature=0x05050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505  
    ephemeral-public-key=0x060606060606060606060606060606060606060606060606060606060606060606  
    record=0x
encoded-packet:  ba4358d5b59616eae9bba87bdf70363fdef69b3ca7006d54713a1bc41c72c465414fc0e2672960eb14b6094c1a38a7d1bd2f2dd6021b4050963d07eb88b6e3c4b8b8c3342690f6d53fe1f09682fb139ad46f5fa2734c2e4eecd835ca726d95e80e5d73d90d022505bc8353749eda90128b32ff00e08687ec40d03fb818736041b7e49d295b4c55f351be13ed6dc5bb5959f5c80e7931c1565e2e925a2c9ea2522611e35e7c358bfc3987df51a2770f0c7b2f7f

# HandshakePacket (flag-2) (WITH ENR)
initiator-key: 0x01010101010101010101010101010101
nonce: 0x020202020202020202020202
source-node-id: 0x0303030303030303030303030303030303030303030303030303030303030303
dest-node-id: 0x0404040404040404040404040404040404040404040404040404040404040404
message: PING: request-id=0x01  enr-sequence-number=0x00
auth-data: HandshakePacket: 
    auth-data-head=(version=0x01 signature_size=0x40 ephemeral_key_size=0x21)  
    id-signature=0x05050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505050505  
    ephemeral-public-key=0x060606060606060606060606060606060606060606060606060606060606060606  
    record=0xf88bb840242198bffec59a15dd134b44a125e39a3f0330aa481ca7359f17dd1889e7427a3f8974a0c4e64cd518619e12ffec0ece01f85e7807d3aa62d07443b847b7566635826964827634826970848d3ca05389736563703235366b31a103e3449d6a29899c3b29b2c4bc27b6407070d7f33b5524fd2e7f1d0d3dfa9bc68483746370829ceb8375647082851d
encoded-packet:  2e47005ae2c0c50eea1f32fd86029a128e26ad45a6e1b5d0dc58759528d237b0ba5bf7d24087e3add447709db01dd1cf53304a64f26e02d6a1dd895932bc2621e60eef1301be21851a04d3dd2b9398d290573cb2baceafe8e5e438953c787832e8295642ed61c676122e98c15e1fe7cb4319ffc72125c63c5bfb760ea945c456d1a919239078d632585ae192184f6e7ddef806d9d2e8530803c07a18597d4631d675ec7815939aea0194d88c608cf4fc253eacb43d8d5cb71a5c722cd835050b7cc606e4d0e9e8b48cb9fd72c48a39c43c268e2307644f13743ba1ce404b68b3588e5b8eff48ae34175e15d5fbcd1ecb0a572abbea4310986a8176b01da5be2a9b42f40740eebd965dd31ec4a0f6acb3e4f9ff935352165ea9fe24dd4357d628df45ce79884255f83e38489c522611e3d9cf00dc787a383be694bef0bb69b456
```

pipermerriam commented on discv5/discv5-wire.md:
```diff
@@ -281,7 +181,33 @@ current record as the only result.
+
+### TALKRESP Response (0x06)
+
+    message-data = [request-id, response]
```
What do you think about adding a `total` field to this message so that it is easier to implement multi-message responses naively.  Without this it appears that value will have to be embedded in the `response` itself which makes it awkward to fit this into the same multi-response framework that the other messages use.

fjl replied:
Hmm, I didn't consider that anyone could want multiple responses for this request.

pipermerriam replied:
My experience with UDP is limited so I may be leaning the wrong way here.  My thinking here is base on:

1. The `Nodes` response message which allows the payload to be distributed across multiple responses.
2. Planning to implement distributed storage of chain data via `TALKREQ/TALKRESP` in which the payloads will be larger than can fit into a single datagram requiring them to be split across multiple messages.

That said, it may be inappropriate to implement that at this level so maybe hold off until I actually get around to using them and i can re-raise this then if it still makes sense.

pipermerriam commented on discv5/discv5-wire.md:
```diff
@@ -263,14 +163,14 @@ PONG is the reply to PING.
 ### FINDNODE Request (0x03)
 
-    message-data = [request-id, distance]
+    message-data = [request-id, distance₁, distance₂, ..., distanceₙ]
```
This would be much easier to implement in our RLP framework as `[request-id, [distance₁, distance₂, ..., distanceₙ]]`

Also has the nice benfit of all of the message formats having a fixed number of elements at their base level.

fjl replied:
I'm OK with it if you must have it this way. Actually, now that you bring it up, we could probably get away without the outer list. That'd be a more radical change though. What do you think?

pipermerriam replied:
> we could probably get away without the outer list. That'd be a more radical change though. What do you think?

I don't understand what you are suggesting here.

fjl replied:
I mean

```
message-data = request-id || [distance₁, distance₂, ..., distanceₙ]
```

pipermerriam commented on discv5/discv5-wire.md:
```diff
@@ -263,14 +163,14 @@ PONG is the reply to PING.
-The node IDs of all nodes in the response must have a shared prefix length of `distance`
-with the recipient's node ID. A request with distance `0` should return the recipient's
-current record as the only result.
+A request with distance `0` should return the recipient's current record as the only result.
```
I think the wording here leaves some ambiguity as to what the correct behavior would be for a request that includes `distance == 0` as well as other distances.  `[<request-id> , 0, 255]`

Logically, it should contain your local ENR **and** some set of ENR records at log distance 255, but the wording here suggests that the response should **only** contain the local ENR record.

fjl replied:
I understand your concern, will update.

pipermerriam commented:
Another implementation tip that may be useful to others.

Consider this case.  There is a handshake in progress between `Initiator` and `Recipient`.  The `Recipient` has sent the `WhoAreYouPacket`  and is waiting for the `HandshakePacket` to finalize the handshake.

The `Recipient` receives a `MessagePacket`.

At this stage there are two possibilities.

- Case A
    - The `MessagePacket` arrived out of order.  A `HandshakePacket` will arrive shortly, after which the `MessagePacket` can be decoded successfully with the derived session keys.
- Case B:
    - The `MessagePacket` is actually the initiation of a new handshake (and there **will not** be a `HandshakePacket` arriving to finalize the current in-progress handshake).

The `Recipient` must make a choice.  

- If the situation is case A they should buffer the packet and decode it in a moment once session keys have been derived.
- If the situation is case B they should initiate a new session and respond with a `WhoAreYouPacket`

Since the `Recipient` doesn't know which case it is, I believe we should encourage implementations to assume case B and to always respond with a `WhoAreYouPacket` when encountering this situation.

- If the situation is actually case A, then the `WhoAreYouPacket` should simply be ignore since the `Initiator` would already consider the handshake completed **and** a `WhoAreYouPacket` will not incidentally trigger initiation of a new handshake.
- If the situation if actually case B, then the `Recipient` will have responded correctly, allowing the new handshake to proceed, with the initial handshake eventually timing-out/expiring.

This is a subtle enough case that it seems potentially worth documenting in some sort of implementers guide since an implementation that chooses the other route will result in both `Initiator` and `Recipient` having incomplete in-progress handshakes that will have to wait to be timed-out/expired before either is likely to initiate again.

fjl commented:
I'm finally updating the spec with the header masking, and two new ideas came up:

- In https://github.com/sigp/lighthouse/issues/1568, it came to light that we need to have a minimum packet size for message packets. I was aware of this issue but didn't think about it lately, mostly because I didn't have a good solution. 

Ultimately, the issue is that WHOAREYOU packets are too large in comparison to message packets. 
This is because they need to carry 32 bytes of unique`id-nonce` random data back to the initiator. 
But now, with the header masking which we are adding, we already send 16 bytes of random `iv` data.
This means we might be able to shrink the `id-nonce` by at least 16 bytes and redefine the ID-nonce hash formula to simply be `sha256(iv || authdata)`.

- Regarding the issue @pipermerriam  described above, I think it might be a good idea to distinguish requests from responses via the `flag` field. One of the underlying assumptions has always been that *responses* (i.e. PONG, NODES, ...) should not initiate a new session. This assumption is in place because the handshake works by resending the request when WHOAREYOU is received as the response. If WHOAREYOU is received as the response to a response, it should be discarded anyway.

If we add another flag value `3` for responses, the recipient would know that it shouldn't respond with WHOAREYOU even if it can't decrypt the message for some reason.

fjl commented:
Damn, it looks like I wiped out the definitions of `id-nonce-input` and `id-signature` when moving the handshake to theory.md. Will put them back!

fjl commented:
Another thing we could still change: feed the dest-id into id-signature.

I'm suggesting this because I got mildly concerned that the removal of the encrypted
auth-response might create an MITM risk. In the 5.0 handshake, we encrypted the
id-signature with a separate session secret, but this was removed and the signature
is now sent clear text.

Consider communication A -> X -> B, with X being the MITM node.

- A has node record of B, i.e. it knows B's ID
- X also has the node record of B

- A sends a message packet (or a random packet).
- Message packet arrives at X, it can unmask the header. X learns about A's ID here.
- X forwards the message packet to B, changing the return address.
- B responds to X with WHOAREYOU containing id-nonce.
- X forwards WHOAREYOU to A.
- A derives session keys and responds to X with handshake packet containing id-signature + eph-pubkey.
- X forwards the handshake packet. X can see signature and eph-pubkey, __but it can't
  change eph-pubkey because it would invalidate id-signature__. X can't decrypt message
  either because it doesn't have any of the private keys.
- B verifies id-signature and derives session keys. B can now decrypt the message.

So, no MITM possible.

But can X replay the id-signature somehow to authenticate as A to a different node C?
Technically not, because X can't influence the id-nonce sent by node C. But it feels too
shaky with all those malleable parts. I think it might be safer to put B's ID into the
signature to ensure this can't happen, if only to make the id-signature an even more
specific statement.

zilm13 commented:
@fjl Do you have Geth fork with all in it?

fjl commented:
@zilm13 my current branch is here: https://github.com/fjl/go-ethereum/tree/p2p-discover-v5-new-format

mkalinin commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+    info             = "discovery v5 key agreement" || node-id-A || node-id-B
+    prk              = HKDF-Extract(secret, id-nonce)
+
+    initiator-key, recipient-key, auth-resp-key = HKDF-Expand(prk, info)
```
Seems like we don't need `auth-resp-key` anymore.

fjl replied:
Yes, that's correct. Good catch!

fjl commented:
@pipermerriam a bit more context about this idea:

> I think it might be a good idea to distinguish requests from responses via the flag field.

When we discussed this on the phone, I wasn't really able to explain why it is generally impossible
to perform the handshake as part of a response, and why this change is desirable. 

Let me try to explain it using a communiction example might help:

```
A <- B   FINDNODE (message packet)

    -- B drops the session here --

A -> B   NODES (message packet)
A <- B   WHOAREYOU
A -> B   NODES (handshake packet)
```

This handshake will not work because B doesn't have a response to send to confirm the session. This means the outcome of this handshake is not verifiable. This is explained in the spec, in the last sentence [of the handshake section](https://github.com/ethereum/devp2p/blob/master/discv5/discv5-wire.md#handshake):

> Node A receives the response and authenticates/decrypts it with the new session keys. If decryption succeeds, node B's identity is verified, A considers the new session keys valid and uses them for all further communication.

Essentially, the session will be confirmed only when the recipient actually **responds** to the handshake packet. But B cannot respond if the handshake packet contains a response message like NODES.

If we want to make this work, we have two options:

- (1) Add a new flag value that says if the message is a response. This will fix it because B will not try to handshake when it gets the NODES message.
- (2) Add a new kind of packet, the 'handshake session confirmation response' which can be sent if session confirmation is required but the node doesn't have any message to send. In the example above, B would need to send this packet as the response to the NODES handshake packet.

I prefer the option (1) because it's simpler and will prevent useless handshakes if the session ends in between a request and response interaction. If both sides use the LRU-based session cache, this is very unlikely to happen anyway.

mkalinin commented on discv5/discv5-wire.md:
```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
+`static-header`, followed by a variable-length `authdata` section (of size `authdata-size`).
 
-At this time, the only supported identity scheme is "v4".
+    header        = type-header || authdata
```
```suggestion
    header        = static-header || authdata
```

mkalinin commented on discv5/discv5-wire.md:
```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
+    authdata      = authdata-head || id-signature || eph-pubkey || record
+    authdata-size = 15 + sig-size + eph-key-size + len(record)
+    authdata-head = version || nonce || sig-size || eph-key-size
+    version       = uint8     -- value: 1
```
What does this version relate to?

mkalinin replied:
I see. What part of the wire protocol could be affected by changing this version?

fjl replied:
Anything after the handshake could change. We can add new messages, change existing messages, or change the handshake crypto.

mkalinin replied:
If it could affect almost everything, doesn't it make sense to add a version to a static header? We have a couple of spare bytes there.

fjl replied:
I don't want to put it into the static header because that would increase the size of all packets.

mkalinin replied:
> I don't want to put it into the static header because that would increase the size of all packets.

Don't we have a couple of unused bytes in `protocol-id` field, which are currently filled with spaces? 

Version does make sense for all packets until handshake is done: `RANDOM_PACKET`, `WHOAREYOU`, `HANDSHAKE`. As layout of each of these packets could be affected by changes in crypto. Having a version in all those packets, also, allows for backward compatibility and smoother upgradability.

fjl replied:
Yes, you're right, we could put it in this place.

fjl commented:
@pipermerriam I'm trying to decode the test vectors you made [here](https://github.com/ethereum/devp2p/pull/157#issuecomment-669289421), but running into trouble. The AES-GCM payload of the PING message packet doesn't decrypt with the provided key "0x01010101010101010101010101010101".

It would be nice if we could use the following secp256k1 private keys for test vectors:

```
node-a-key: eef77acb6c6a6eebc5b363a475ac583ec7eccdb42b6481424c60f59aa326547f
node-b-key: 66fb62bfbd66b9177a138c1e5cddbe4f7c30c343e94e68df8769459cb1cde628
```

The node IDs (with the 'v4' scheme) corresponding to these keys are:

```
node-a: aaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
node-b: bbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
```

Defining the private keys is important because it will allow us also make test vectors with valid signatures on the node record, for example.

mkalinin commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+The number of messages which can be encrypted with a certain session key is limited
+because encryption of each message requires a unique nonce for AES-GCM. In addition to the
+keys, the session cache must also keep track of the count of outgoing messages to ensure
+the uniqueness of nonce values. Since the wire protocol uses 96 bit AES-GCM nonces, it is
```
According to [this recommendation, section 8.2.2](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf#%5B%7B%22num%22%3A65%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C792%2Cnull%5D), it would be secure to get the first `IV` from a randomly generated bit string and then increment its e.g. 32-bit part to derive values for next `IVs`.

fjl replied:
Yes, that would work as well.

mkalinin replied:
Sure, I was also trying to say that 32 bit counter + 64 random bits is not considered as a secure option. It might be worth changing this sentence to what we have in the recommendation.

pipermerriam commented:
@fjl 

Regarding https://github.com/ethereum/devp2p/pull/157#issuecomment-685517761

IIUC the problem with allowing *response* messages trigger handshake initiations is that according the the protocol the initiator will not consider the handshake to have been successful until they have received and decode a message using the derived session keys.

It seems we don't need to do anything to remedy this situation because either:

1. Some *other* message will be sent by the recipient using the new keys, allowing the initiator to validate the keys are correct.
2. No other messages will pass between the two nodes and the session will eventually be evicted.
3. The initiator will eventually send a message to the initiator, and upon receiving the response to that second message, the initiator can consider the session keys valid.

It seems the spec-as is, but maybe with a bit more narrative explanation of this exact situation can deal with this situation just fine.

pipermerriam commented:
@fjl I think that my implementation needs to be updated to latest spec, after which I'll re-generate new test vectors (with the keys you recommended).

kdeme commented on discv5/discv5-wire.md:
```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
-At this time, the only supported identity scheme is "v4".
+    header        = static-header || authdata
+    static-header = protocol-id || src-id || flag || authdata-size
+    message       = aesgcm_encrypt(initiator-key, nonce, message-plaintext, header || authdata)
```
```suggestion
    message       = aesgcm_encrypt(initiator-key, nonce, message-plaintext, header)
```
Is `header || authdata` the intention here?
I'd expect `header` or `static-header || authdata`

AgeManning replied:
I think `header` should be changed to `masked-header` and the `ad` part of the aesgcm_encrypt should be `masked-header`

kdeme replied:
Additionally, the `authdata` can get rather big in case of a Handshake Message Packet. And while that should work, it seems unnecessary?

fjl replied:
You're right, the intention was `static-header || authdata`. I just forgot to update it when renaming the header.

The purpose of this authenticated data input is binding the message to the packet it is contained in. Processing of this data is very fast and I would prefer to put in as much of the packet as possible here.

kdeme replied:
Thanks for the context on why. 

I understand you want to provide as much data as possible but I find it still weird because of the (big) difference of ad size between the two packet types, especially as you would be sending more often the `Ordinary Message Packet`. But ok, it does no harm I guess.

fjl commented:
Here's the list of things that still need a decision or additional spec text:

- [ ] dest-id in id-signature input: https://github.com/ethereum/devp2p/pull/157#issuecomment-683719203
- [ ] require WHOAREYOU response when message package received during handshake: https://github.com/ethereum/devp2p/pull/157#issuecomment-679481919
- [ ] add flag value for distinguishing requests and responses: https://github.com/ethereum/devp2p/pull/157#issuecomment-685517761
- [ ] solution for mininum packet size: https://github.com/ethereum/devp2p/pull/157#issuecomment-683698160

Nashatyrev commented on discv5/discv5-wire.md:
```diff
@@ -19,6 +18,9 @@ Here we present the notation that is used throughout this document.
     means binary XOR of `a` and `b`\
 `sha256(x)`\
     is the SHA256 digest of `x`\
```
`rlp_bytes(x)` (above) doesn't look to be used anymore. M.b. remove it?

fjl replied:
Thanks, I will remove!

Nashatyrev commented on discv5/discv5-wire.md:
```diff
@@ -49,187 +51,107 @@ request/response and 1s for the handshake.
-    token            = auth-tag of request
-    id-nonce         = 32 random bytes
-    enr-seq          = highest ENR sequence number of node A known on node B's side
+![whoareyou packet layout](./img/whoareyou-packet-layout.png)
```
On the diagram there is a capture `AESGCM nonce or message packet` 
What the `... or message packet` stands for?

fjl replied:
It says 'AESGCM nonce **of** message packet', but maybe it's hard to read because of the color.

Nashatyrev replied:
Oh right, sorry!

pipermerriam commented:
@fjl my opinions

- dest-id in id-signature input:
    - no strong opinion.  it sounds like it does mitigate *something* and that the cost of including it is very low.
- require WHOAREYOU response when message package received during handshake:
    - This was my suggestion.  I think as a **SHOULD** style recommendation in an implementation guide would be adequate.  Clients that don't implement this wouldn't cause problems in any meaningful way.
- add flag value for distinguishing requests and responses:
    - I'm :-1: on this.  I don't see any problems with having a handshake be initiatable via a response message.  My full thoughts can be found in [this comment](https://github.com/ethereum/devp2p/pull/157#issuecomment-685823255)
- solution for mininum packet size:
    - I'm :+1: on the proposed approach to reduce the WHOAREYOU message size by reducing the size of the `id-nonce` and including the 16 bytes of `iv` data.  This gets the packet size reduced by 16 bytes which I believe leaves us with a difference of 24 bytes between the base size of the two packets.  Looking at the various packets we send, that likely means 10-20 bytes of padding for PING and FIND_NODES messages which have minimal payload sizes.  This seems acceptable.

Nashatyrev commented:
@pipermerriam are those test vectors up-to-date? https://github.com/ethereum/devp2p/pull/157#issuecomment-669289421

fjl commented:
Here are some new test vectors matching the current spec text.
These are VALID packets, i.e. they could actually be sent on the wire.

WHOAREYOU packet (flag 1):
```
# src-node-id = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
# dest-node-id = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
# whoareyou.auth-tag = 0x0102030405060708090a0b0c
# whoareyou.id-nonce = 0x0000000000000000000000000000000000000000000000000000000000000000
# whoareyou.enr-seq = 0
# 

7899173fcecaf651504b79c3bb5b930dd99c619153f3c93f9afcecb214f45543
b4e70b4fc8748887616eb793f2e90b19cc3a070fe957afdc3a639ca891144bfe
63dee066f5a78096bcf38d382b2f4a434d73986e0e4e018759a66b389a7d3f95
9edba38ebb03851c3e31f3e11f32ae
```

Ping message packet (flag 0):
```
# src-node-id = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
# dest-node-id = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
# read-key = 0x00000000000000000000000000000000
# auth-tag = 0xffffffffffffffffffffffff
# ping.req-id = 0x00000001
# ping.enr-seq = 2

4076162ea9513c747a2783efb1be1030ca1e91aaf871960fc9f477fd195bc2a4
1a89c03f7c519fe0275cabbf9875874ca66e6a6908776d796a8ec9ffb6de03fc
995f10628cbdf0b84102ed931f66d180cbb4219f369a24f4e6b24d7bdc2a04
```

Ping handshake packet (flag 2):
```
# src-node-id = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
# dest-node-id = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
# whoareyou.auth-tag = 0x0102030405060708090a0b0c
# whoareyou.id-nonce = 0x0102030405060708090a0b0c0d0e0f1000000000000000000000000000000000
# whoareyou.enr-seq = 1
# auth-tag = 0xffffffffffffffffffffffff
# ping.req-id = 0x00000001
# ping.enr-seq = 2

147ebae1f23aff4db44aecc4d653311c64ae2ad184fda5d5865760816c707256
e779a21f56d9ff2f0b084d39a466cf9287bee46eb20a6e6b6426cfd6f3b14de5
b776fd7341150d129586a3cb9ae59694986c7f513ccb821a17882634eec434da
c1ee0af96b60546e0f9f7a0e4ab7d89783370fe74fbd81870c0e15338772ec5d
423bb65dd50ecf810dffab6c22e1365c2d0d3323ed3a718e8e82ab7c857d7240
c90dd92efcdbd11f5a1d8ce05961f9ea0d899a16ea973bd08b527dfdcb49be01
2c84d2e40daca630076629f01d9e04a987312b748c74fb5da7a524fd8fbf32cb
e249
```

Ping handshake message packet (flag 2, with ENR):
```
# src-node-id = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
# dest-node-id = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
# whoareyou.auth-tag = 0x0102030405060708090a0b0c
# whoareyou.id-nonce = 0x0102030405060708090a0b0c0d0e0f1000000000000000000000000000000000
# whoareyou.enr-seq = 0
# auth-tag = 0xffffffffffffffffffffffff
# ping.req-id = 0x00000001
# ping.enr-seq = 2

d85d60d07419d333acefdbc8e9c5fa1251fd4b32884a2290f764063260a68308
5d89dd33a5893dc32a877ff47f4980d4367dfe062d3182891cd5392519846a06
5052d3578d1756b199af4483993e6ada62f60c7c330dd905c1a55e6907a6ee8f
4c72667e4828d8c586dea5f2734ad5495945ce96c97e30f58dc7a91268435732
7fb9bc68d156bd9888a21c13a91fbf446f348f1cfbba53d935c97a820818a099
cab323c587a037031bf6cdf2c5a90ecff53b5092994f5aabb370a98a820631fa
e0c230913219b814e1198c702705d4363b9a67d70f7228a0bd2222d07d637187
fe5efbafb6f11eb3726b3d3fb75261aec0ab1fdd88eddbd564a5a9d78e5b3fad
d975fe57adc6251c1a9530a3fed1c6215c5e09b8cca0fecb234bd6669200c999
cd19f311595a69562e3f7d93da096247d9f42a4e50ec156362dcf80033531e4a
6c34d98f185c18ea25d26e2c47732b4cc2ec6317ea746e9a3491d66a30993067
f1
```

Nashatyrev commented:
May be it makes sense to limit the message `req-id` field size? I would suggest max 8 bytes

fjl commented:
Yes that would make sense.

Nashatyrev commented:
> Here are some new test vectors matching the current spec text.

Hey @fjl! 
Seems like the last two vectors descriptions are not correct: these are `handshake` packets while description contains `whoareyou`. I've got the following data decoding those messages: 
```
PingMessage{requestId=0x00000001, enrSeq=1}
Header{
	sourcePeerId=0xAAAA8419E9F49D0083561B48287DF592939A8D19947D8C0EF88F2A4856A69FBB, 
	authData=HandshakeAuthData{
		version=1, 
		aesGcmNonce=0xFFFFFFFFFFFFFFFFFFFFFFFF, 
		idSignature=0xC14A44C1E56C122877E65606AD2CE92D1AD6E13E946D4CE0673B90E237BDD05C2181FC714C008686A08EB4DF52FAAB7614A469576E9AB1363377A7DE100AEDC2, 
		ephemeralPubKey=0x9A003BA6517B473FA0CD74AEFE99DADFDB34627F90FEC6362DF85803908F53A50F497889E4A9C74F48321875F8601EC65650FA0922FDA04D69089B79AF7F5533, 
		enrBytes=0x
	}
}
		
PingMessage{requestId=0x00000001, enrSeq=1}
Header{
	sourcePeerId=0xAAAA8419E9F49D0083561B48287DF592939A8D19947D8C0EF88F2A4856A69FBB, 
	authData=HandshakeAuthData{
		version=1, 
		aesGcmNonce=0xFFFFFFFFFFFFFFFFFFFFFFFF, 
		idSignature=0xC14A44C1E56C122877E65606AD2CE92D1AD6E13E946D4CE0673B90E237BDD05C2181FC714C008686A08EB4DF52FAAB7614A469576E9AB1363377A7DE100AEDC2, 
		ephemeralPubKey=0x9A003BA6517B473FA0CD74AEFE99DADFDB34627F90FEC6362DF85803908F53A50F497889E4A9C74F48321875F8601EC65650FA0922FDA04D69089B79AF7F5533, 
		enrRecord=NodeRecordV4{
			publicKey=0x0313D14211E0287B2361A1615890A9B5212080546D0A257AE4CFF96CF534992CB9, 
			udpAddress=empty, 
			nodeId=0xAAAA8419E9F49D0083561B48287DF592939A8D19947D8C0EF88F2A4856A69FBB
		}
	}
}
```

mkalinin commented:
- dest-id in id-signature input:
  If we want we can use `sha256("discovery-id-nonce" || initiator-key)` for `id_sign`. This enshrines a session identity into the signature.

- require WHOAREYOU response when message package received during handshake:
  > - If the situation is actually case A, then the WhoAreYouPacket should simply be ignore since the Initiator would already consider the handshake completed and a WhoAreYouPacket will not incidentally trigger initiation of a new handshake.

  Will recipient resend the same WHOAREYOU or create a new one?

- add flag value for distinguishing requests and responses:
  I agree with @pipermerriam that there is no big problem with handshakes initiated by responses, my vote is for simplicity because this case looks too rare to be specially addressed.

- solution for mininum packet size:
  > and redefine the ID-nonce hash formula to be sha256("discovery-id-nonce" || iv || authdata || eph-pubkey). Is this too risky? Need to figure it out!

  Why would we need to redefine the formula? I don't see any risk in replacing `id-nonce` with `iv` and shrinking it down to 16 bytes. From KDF standpoint this change does look safe.

fjl commented:
> If we want we can use sha256("discovery-id-nonce" || initiator-key) for id_sign. This enshrines a session identity into the signature.

I don't want to do this because then we would need to run the key derivation before checking the signature. It increases the computational cost of verifying invalid handshakes.

> Why would we need to redefine the formula? I don't see any risk in replacing id-nonce with iv and shrinking it down to 16 bytes. From KDF standpoint this change does look safe.

Yes, I've come to the same conclusion, will update it.

fjl commented:
Looks like there is a bug in my test vectors: the ephemeral pubkey is 65 bytes, but should be 33 bytes (compressed key). I will update them once I have the other spec changes in as well.

kdeme commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+controls the identity key which signed the record and also prevents replay of the
+handshake.
+
+    id-nonce-input   = sha256("discovery-id-nonce" || id-nonce || ephemeral-pubkey)
```
The `ephemeral-pubkey` that goes over the wire now is in the compressed format (33 bytes).
Do we still used the uncompressed format when creating the id-nonce-input as it used to be? Or directly the compressed format?

fjl replied:
It should be the compressed format for the id signature as well.

fjl replied:
Not sure if this needs additional text or if it's fine this way.

mkalinin commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+controls the identity key which signed the record and also prevents replay of the
+handshake.
+
+    id-nonce-input   = sha256("discovery-id-nonce" || id-nonce || ephemeral-pubkey)
```
Assuming that `id-nonce` is a random value and `ephemeral-pubkey` is highly likely a random value, I am wondering does `"discovery-id-nonce"` constant give any valuable input to this `sha256` computation?

fjl replied:
It binds the signature to the specific purpose it is used for. Signatures are valuable and it is good practice to ensure that the input data is 'qualified', i.e. the input data says what the signature is supposed to mean. You can find many examples of this practice in other protocols as well.

fjl replied:
For example, this kind of thing is also used in TLS 1.3: https://www.davidwong.fr/tls13/#section-4.4.3

> The context string for a server signature is "TLS 1.3, server CertificateVerify". The context string for a client signature is "TLS 1.3, client CertificateVerify". It is used to provide separation between signatures made in different contexts, helping against potential cross-protocol attacks.

mkalinin replied:
Thanks for clarification. It makes sense in their context cause certificates are scarce resource wrt random values obtained from secure RNG, and different protocols may use the same certificates or keys. Anyway, following best practices does make a lot of sense. 

Do we want to mix a version, i.e. `1` in our case, into the context as well?

mkalinin replied:
Also, why not sign `id-nonce-input` with `ephemeral-secret-key` as `ephemeral-pubkey` goes alongside with the signature? In this case we could replace `ephemeral-pubkey` with `dest-node-id` for free which is probably a desirable thing.

fjl replied:
I don't understand the suggestion. The purpose of the ID signature is to prove that the initiator node actually has the private key that signed its ENR. The identity key of the initiator isn't used anywhere else in the handshake, so if we signed the ID nonce with the ephemeral key, the identity wouldn't be proven.

mkalinin replied:
Yes, you're right. The identity check is what I've been missing.

Nashatyrev commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+
+Node A must have a node record for node B and know B's node ID to communicate with it. If
+node A has session keys from prior communication, it encrypts its request with those keys.
+If no keys are known, it initiates the handshake by sending a packet with random content.
```
Is there any minimal random message size requirement as in v5.0?

fjl replied:
I'm still working on it.

fjl replied:
This is added now.

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -281,7 +210,33 @@ current record as the only result.
+### TALKREQ Request (0x05)
+
+    message-data = [request-id, protocol, request]
+    message-type = 0x09
```
```suggestion
    message-type = 0x05
```

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -281,7 +210,33 @@ current record as the only result.
+
+### TALKREQ Request (0x05)
+
+    message-data = [request-id, protocol, request]
```
Can we specify the types of `protocol` and `request` also.

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -281,7 +210,33 @@ current record as the only result.
+### TALKRESP Response (0x06)
+
+    message-data = [request-id, response]
+    message-type = 0x10
```
```suggestion
    message-type = 0x06
```

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -281,7 +210,33 @@ current record as the only result.
+
+### TALKRESP Response (0x06)
+
+    message-data = [request-id, response]
```
Can we specify the type of response

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -297,7 +252,10 @@ REGTOPIC is always answered by a TICKET response. The requesting node may also r
+**Implementations should not respond to or send these messages.**
 
     message-data = [request-id, ticket, wait-time]
     message-type = 0x06
```
```suggestion
    message-type = 0x08
```

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -308,7 +266,10 @@ TICKET is the response to REGTOPIC. It contains a ticket which can be used to re
+**Implementations should not respond to or send these messages.**
 
     message-data = [request-id, topic]
     message-type = 0x07
```
```suggestion
    message-type = 0x09
```

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -318,7 +279,10 @@ REGCONFIRMATION notifies the recipient about a successful registration for the g
+**Implementations should not respond to or send these messages.**
 
     message-data = [request-id, topic]
     message-type = 0x07
```
```suggestion
    message-type = 0x10
```

AgeManning commented on discv5/discv5-wire.md:
```diff
@@ -318,7 +279,10 @@ REGCONFIRMATION notifies the recipient about a successful registration for the g
 registration has elapsed on a topic queue.
 
-### TOPICQUERY Request (0x08)
+### TOPICQUERY Request (0x10)
```
If this is hex, shouldn't it make more sense to `0x0a` ?

fjl replied:
Yes 🤦

mkalinin commented on discv5/discv5-wire.md:
```diff
@@ -49,187 +49,113 @@ request/response and 1s for the handshake.
-(`auth-tag`).
+    authdata      = authdata-head || id-signature || eph-pubkey || record
+    authdata-size = 15 + sig-size + eph-key-size + len(record)
+    authdata-head = version || nonce || sig-size || eph-key-size
```
`authdata-head` should also contain plain `id-nonce` string. Making `id-nonce` validation through signature verification opens a DoS vector. An adversary will be able to create handshake packet once and force remote peers to verify signatures consuming their CPU.

A particular attack scenario outlined by @Nashatyrev: malicious peer sends a couple of packets with a small timeout between them, first one is treated as a random packet and initiates the handshake process, second packet is the handshake packet which forces honest peer to verify signature before it's able to recognise an invalid `id-nonce`. In this attack the same message can be used for both packets, it can be reused between different targets and attack rounds.

pipermerriam replied:
I don't understand how including `id-nonce` here prevents this.  It seems that you could still issue a bogus handshake response with a bogus signature in this case and that the bogus signature can still be re-used across different target attacks.  It's unclear to me what including the `id-nonce` here does to prevent that.

mkalinin replied:
Sure, you can, but if `id-nonce` is not included it makes this attack much more efficient. An attacker will have to wait for the `WHOAREYOU` message to include valid `id-nonce` to be able to pass an initial check and trigger signature verification.

We can't completely eliminate the DoS vector, ratelimiting seems to be the only way to mitigate it.

fjl replied:
In general, the effectiveness of this attack also depends on the performance of signature verification. In discv4, we have a signature on every packet, and this attack was never considered because signature verification is fast. Sure, it might be possible to reject the bad packet even faster if including the nonce is required, but it will also increase the size of the handshake packet. We need to weigh these factors. I'm personally not to worried about this particular attack
because I think signature verification is fast enough.

pipermerriam replied:
The only thing the proposed solution seems to fix is that it forces the packet to be sent **after** the WHOAREYOU meaning that in both situations one must still do signature verification so it seems that the speed of signature verification doesn't actually matter.

It seems that all that matters here is whether the increased cost of the attacker having to wait for the WHOAREYOU packet prior to sending the bogus HANDSHAKE packet makes a meaningful difference.  In both cases the recipient will have to perform signature verification.  This leads me to think that this may be better addressed via some form of artificial rate limiting on inbound handshakes.

My opinion is that we should not make the proposed protocol change but I'm open to understanding where my reasoning is off if someone sees something here that I'm missing.

mkalinin replied:
> In discv4, we have a signature on every packet, and this attack was never considered because signature verification is fast. 

That is true, signature verification takes ~1ms on my laptop with core i7 CPU.

I have the following concern. Relatively cheap [spoofed VPS](https://www.clientvps.com/service/spoofed-vps) could turn this DoS vector into, effectively, DDoS  vector, i.e. one server becomes able to attack from tons of remote peers triggering tons of signature verifications per unit of time. With `id-nonce` in place an attacker will have to read `WHOAREYOU` message which are sent to IPs that he has no ownership over hence a single machine attacker will be able to trigger sig verification for only one handshake at a time.

> This leads me to think that this may be better addressed via some form of artificial rate limiting on inbound handshakes.

Definitely. Artificial rate limiting is the protection that we want in the first place. And it might also be enough to mitigate that DoS leveraged by spoofing.

> but it will also increase the size of the handshake packet

It's gonna be 16 bytes (with `IV` replacing `id-nonce`) per session. Is this a significant increase for the handshake packet?

pipermerriam replied:
I think I'm +0 on this having thought about it more.  Requiring the `id-nonce` value to be present in the HANDSHAKE packet seems to provide at least *some* increased cost to executing the attack.

The reason I'm not sold is this.  Even with the change, an attacker can just include a random value.  While a proper implementation can avoid the extra signature verification since the `id-nonce` check will fail to match, the attacker can still force the initial message packet to be parsed, the WHOAREYOU packet to be generated and sent, and the HANDSHAKE packet to be parsed and `id-nonce` checked.  The **only** thing this seems to do is *maybe* save one signature verification which we've already established is inexpensive... 

I do however see that this is a small change, and the cost of including the `id-nonce` in the HANDSHAKE header is indeed only 16 bytes and it does seem to provide *some* improvement.  I'm good with whatever decision is made on this (which I'm :+1: on letting @fjl make the final call).

fjl replied:
I'm still thinking about it.

fjl replied:
I have decided to not address it in v5.1. This is something we can change in v5.2.

fjl replied:
I just realized there is a good way to address this without increasing the handshake packet size even more: just replace the src-id with id-nonce.

mkalinin replied:
> I just realized there is a good way to address this without increasing the handshake packet size even more: just replace the src-id with id-nonce. We can definitely consider it for v5.2.

I can file an issue with this particular proposal. So we won't forget to consider it during further update.

fjl replied:
Yes, please do that!

fjl commented on discv5/discv5-theory.md:
```diff
@@ -41,6 +40,141 @@ IP address and port.
+    A <- B   WHOAREYOU including id-nonce, enr-seq
+
+Node A now knows that node B is alive and is ready to perform the handshake. The handshake
+proceeds by re-sending the original request message in a [handshake packet].
```
I've tried to address this with improved text now. The general idea with the handshake is that you're always trying to send a message packet (otherwise you wouldn't be sending anything). But, when you don't have keys so encrypt it with, you send a random packet because that's the same as encrypting the message with a random key. When you get the WHOAREYOU, you just send the message again, but in a handshake message packet.

Nashatyrev commented on discv5/discv5-wire.md:
```diff
@@ -263,14 +183,17 @@ PONG is the reply to PING.
-current record as the only result.
+The recipient should create the result set by collecting nodes from its local node table
+according to the requested distances. Implementations should limit the number of nodes in
+the result set. The recommended result limit for FINDNODE queries is 16 nodes.
```
May be make sense to settle some explicit limit on the total returned ENRs count for a single request (across all response packets)? The doc says '_should_ limit' but doesn't say what the limit _must_ be
I.e. I'm not sure if I should treat 100K ENRs returned as a remote node misbehavior or not.

fjl replied:
I've added something for this. I don't want to be to too specific about this yet, but it's definitely a good idea to limit it in the implementation. In go-ethereum, the limit is max 5 responses.

mkalinin commented:
We propose to include a couple of recommendations that nodes may implement to mitigate DoS and eclipse attacks. Counter-measures are described here https://github.com/ethereum/devp2p/issues/161

cc @jrhea

fjl commented:
I've just pushed the last spec update. Thank you so much for all your amazing feedback on this pull request. I have tried to address all of the issues you guys have raised.

Here's the list of changes in the last couple commits:

- The `version` field of the handshake message packet is gone. Instead, the `protocol-id` is now
  6 bytes (it was 8 before) and the two bytes after `protocol-id` are now the version.
- `id-nonce` size is now 16 bytes
- The definitions for `id-signature` and the key derivation have been changed to include
  the `masking-iv` and some data. Please check the definitions against your implementation
  to see what you need to change. I will publish new test vectors for this part very soon.
- The minimum packet size is now **90 bytes** for all packets. The size of WHOAREYOU is
  95 bytes. It doesn't fully solve the amplification issue, but I think it's acceptable. There are
  other protocols with worse amplification factors and I don't want to add padding to message
  packets to compensate for this difference.

We need to stop changing the spec now and focus on interop.

kdeme commented on discv5/discv5-wire-test-vectors.md:
```diff
@@ -1,339 +1,159 @@
+    node-id-a = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
+    node-id-b = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
+    whoareyou.masking-iv = 0x01010101010101010101010101010101
+    whoareyou.id-nonce = 0x02020202020202020202020202020202
```
id-nonce is provided here, but the spec says:
`challenge-salt     = masking-iv || authdata`
and:
`prk                = HKDF-Extract(secret, challenge-salt)`

So I'd expect `authdata`. But I'm guessing it got perhaps swapped in the spec?

pipermerriam replied:
I believe `authdata` can be derived from serializing the WHOAREYOU packet.

kdeme replied:
Yeah, the problem is not about where to get the authdata or the idnonce from. It is rather that currently the spec and the test vectors seem to have have swapped around of where to use `challenge-salt` and where to use `challenge-id-input`.

pipermerriam commented on discv5/discv5-wire.md:
```diff
@@ -100,7 +100,7 @@ In ordinary message packets and handshake message packets, the packet contains a
 empty. Implementations must generate a unique `nonce` value for every packet.
 
-    message       = aesgcm_encrypt(initiator-key, nonce, message-pt, header)
+    message       = aesgcm_encrypt(initiator-key, nonce, message-pt)
```
I was curious about this too.

Nashatyrev replied:
Assuming mitm attack: 
- for `ordinary` packet the only field which may be tweaked is `version`. Now there is nothing to do with it. But another version value would be possible there could be interesting cases
- for `handshake` packet almost all fields are signed with `id-signature` but there are still fields which may be tweaked. No ideas on a possible attack for now, but this change looks as security weakening

zilm13 replied:
@Nashatyrev for `handshake` MITM could swap ENR for an old version with an old host for example. Not clear what to do with packets on it after this without a private key, but looks not good.

fjl replied:
Hey, thanks for this analysis. I removed the authenticated data because seemed kind-of useless for regular message packets, but we can totally put it back!

Nashatyrev replied:
I'd vote for bringing it back cause without authorized header it looks shaky

pipermerriam commented on discv5/discv5-wire-test-vectors.md:
```diff
@@ -1,339 +1,159 @@
-#### Input Parameters
+    static-key = 0xfb757dc581730490a1d7a00deea65e9b1936924caaea8f44d476014856b68736
+    masking-iv = 0x01010101010101010101010101010101
+    authdata = 0x0102030405060708090a0b0c0d0e0f100000000000000000
```
the `authdata` can be decoded as a WHOAREYOU packet to recover the `id-nonce`

kdeme replied:
Yeah, sure. It is just confusing to provide this instead of immediately id-nonce. But I guess it can be useful for some implementations.

pipermerriam commented on discv5/discv5-wire-test-vectors.md:
```diff
@@ -1,339 +1,159 @@
-#### Expected Outputs
-
-    id-nonce-sig: 0xc5036e702a79902ad8aa147dabfe3958b523fd6fa36cc78e2889b912d682d8d35fdea142e141f690736d86f50b39746ba2d2fc510b46f82ee08f08fd55d133a4
+    id-signature = 0xd82364cfffb18101355371de84ee0def3dca31191b9add79b21a14f4442b6df02dc26df6278f71c83d43645da13071881cacdb43b0aea1e256cdec73a73faf01
```
@fjl I'm not able to reproduce this.  Can you maybe 1) check it against your implementation and 2) verify some of the intermediate values.

```
id-signature-input = 0x646973636f76657279207635206964656e746974792070726f6f66010101010101010101010101010101010102030405060708090a0b0c0d0e0f10039961e4c2356d61bedb83052c115d311acb3a96f5777296dcf297351130266231bbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
message-hash: 0x762d64d468d019e113ba19dd68d4e11be16d989332e45402026871918b08d3f8
signature: 0x60e8ef785881dd89758b26346db5c0a779182d011519ecbb05aed14466db0a4a04b3853e3e34bec6517021c6e6d7beb293f3ba25305c80d094105e17849b0315
```

kdeme replied:
I too cannot reproduce the value.
I can reproduce your values however, but only if I use just the `id-once` field in the `authdata`, which would be according to spec of course.

kdeme commented on discv5/discv5-wire-test-vectors.md:
```diff
@@ -1,339 +1,159 @@
     encryption-key: 0x9f2d77db7004bf8a1a85107ac686990b
     nonce: 0x27b5af763c446acd2749fe8e
     pt: 0x01c20101
     ad: 0x93a7400fa0d6a694ebc24d5cf570f65d04215b6ac00757875e3f3a5f42107903
```
Just wanted to note that `ad` is still used here in the test vector.

fjl commented:
Seeing all this confusion about the challenge fields and the concerns about packet header authentication, I have decided to use the simplest solution possible: just use the entire unmasked packet header as the challenge data for both id signature and KDF, and also use it as authenticated data in the AES/GCM step.

From a spec perspective, this change is very nice and I wonder why we didn't do it like that earlier.

From an implementation point of view, this change sucks, at least for my own implementation. Since the
data dependencies have changed so much, I had to change the ordering of some of the encoding steps
in order to be able to have the encoded headers available when encrypting the message. But it's done in
go-ethereum now, and the test vectors are also updated.

BTW, I noticed that there was a problem with my HKDF usage, my own tests worked on Go 1.15 but not
on earlier versions. This is now corrected and I hope you will be able to match the new test vectors.

kdeme commented on discv5/discv5-wire-test-vectors.md:
```diff
@@ -1,339 +1,156 @@
+    dest-pubkey = 0x0317931e6e0840220642f230037d285d122bc59063221ef3226b1f403ddc69ca91
+    node-id-a = 0xaaaa8419e9f49d0083561b48287df592939a8d19947d8c0ef88f2a4856a69fbb
+    node-id-b = 0xbbbb9d047f0488c0b5a93c1c3f2d8bafc7c8ff337024a55434a0d0555de64db9
+    whoareyou.masking-iv = 0x01010101010101010101010101010101
```
Forgot to update these?
I managed to match your results with instead:
 `challenge-data = 0x000000000000000000000000000000006469736376350001010102030405060708090a0b0c00180102030405060708090a0b0c0d0e0f100000000000000000`

fjl replied:
Thanks! Yes, I pasted in the new results and wanted to delete the old text but deleted the new text instead.

kdeme commented:
> BTW, I noticed that there was a problem with my HKDF usage, my own tests worked on Go 1.15 but not on earlier versions. This is now corrected and I hope you will be able to match the new test vectors.

I can confirm that I can now decode your latest provided test vectors :+1:

fjl commented:
Very nice!

pipermerriam commented:
I'm now passing all of the test vectors in my implementation.

Nashatyrev commented:
I'm passing them too with JVM implementation
